implicits/src/com/rallyhealth/weepickle/v1/implicits/MacroImplicits.scala

```diff
@@ -333,6 +333,139 @@ object MacroImplicits {
           q"if ((found & (1L << ${arg.i})) == 0) {found |= (1L << ${arg.i}); storeAggregatedValue(${arg.i}, ${arg.default})}"
       }"""
 
+    /*
+     * Original pattern-match based method (not used, here more for documentation)
+     */
+    def getKeyIndexUsingPatternMatch(args: Seq[Argument]): Tree =
+      q"""${c.prefix}.objectAttributeKeyReadMap(s.toString).toString match {
+                case ..${
+        for (arg <- args) yield cq"${arg.mapped} => ${arg.i}"
+      }
+                case _ => -1
+      }"""
+
+
+    /*
+     * https://en.wikipedia.org/wiki/Radix_tree -- performance is better than just pattern matching by 20% - 60%
+     * depending of the size of the case class. Fewer gains for case classes with 5 - 8 fields, more gains for
+     * (weirdly) both smaller and larger case classes.
+     *
+     * Assigns currentIndex within visitKeyValue using the key value length as the primary driver. Within each
+     * length, Radix logic compares common string prefixes only once. In fact, any character will only be compared
+     * with a positive match at most once -- there may be multiple mismatches per character, but these are minimized
+     * by having logic in regionMatches to return mismatch at the first mismatching character in a region.
+     *
+     * For example, the following shows the matching logic for fields "icn" (only three-character field), "claimFacility"
+     * and "claimRemarkCd" (same length with common prefix "claim"), "icnSuffixCode" and "icnVersionNum" (also the same
+     * length with common prefix "icn"), and "claimLevelAdjudicationAmt" (only 25 character field).
+     *
+     *  val cs = WeePickle.objectAttributeKeyReadMap(s match {
+     *     case alreadyCs: CharSequence => alreadyCs
+     *     case force => force.toString
```

Comment:
This is for msgpack's non-`visitString` keys, right?

Reply:
Ah, `def visitKey() = com.rallyhealth.weepickle.v1.core.StringVisitor` is still in place. Case classes only ever have string field names. This is intended as a conservatively safe `.asInstanceOf[CharSequence]`.

I wonder if the straight cast would perform slightly better.

Reply:
Yes, this is intended as a conservatively safe/near zero-cost `.asInstanceOf[CharSequence]`. Since the signature is `Any`, not really positive what will be thrown at it. This goes back to the age-old question of what _can_ a key be? Lots of places in WeePickle assume it is a `String`, though not all. Perhaps the signature of `visitKeyValue` in `Visitor` should be changed such that the parameter is guaranteed to be at least `CharSequence`? (Much larger refactor.)

Reply:
I can't find a difference in perf between the two. Nevermind!
```
[info] Benchmark             Mode  Cnt  Score   Error  Units
[info] MatchBench.withCast   avgt    3  3.528 ± 0.679  ns/op
[info] MatchBench.withMatch  avgt    3  3.386 ± 0.621  ns/op
```

Reply:
Slightly surprising -- I wonder if the JIT manages to optimize this into irrelevance...